src/hardware/pwm.rs

```diff
@@ -183,42 +138,42 @@ impl Pwm {
         }
 
         fn set_pwm<P: PwmPin<Duty = u16>>(pin: &mut P, duty: f32) -> Result<(), Error> {
-            let max = pin.get_max_duty() as f32;
-            let value = duty * max;
-            if !(0.0..max).contains(&value) {
+            let max = pin.get_max_duty();
```

Comment:
Is this inclusive or exclusive (like dac MAX_CODE and `adc.max_sample()`)?

Reply:
Fixed now. Note to self:
SI unit `x` to machine unit `y` conversion should follow the following algorithm:

```rust
const MIN: i32 = 0;  // inclusive lower limit
const MAX: i32 = 1 << 18;  // exclusive upper limit

let z = x * SCALE + OFFSET;

// conversion rounds and clamps
// i32 intentionally strictly larger range than the output type
let y = z as i32;

// do bounds check after rounding, clamping and on a datatype with strictly larger range than the output datatype
if !(MIN..MAX).contains(&y) {
    return Err(Error::Bounds);
}

// convert to final output data type here
y as u16
```

Reply:
I see, so do rounding to a bigger int type before bounds checking and then go to final size after.

Reply:
At least that's what I think is the best approach.